core/src/datayoga_core/blocks/relational/write/block.py

```diff
@@ -19,40 +21,40 @@ class OpCode(Enum):
     UPDATE = "u"
 
 
-def get_field(item: Union[Dict[str, str], str]) -> str:
-    return str(next(iter(item.values()))) if isinstance(item, dict) else item
+def get_source_column(item: Union[Dict[str, str], str]) -> str:
+    return next(iter(item.values())) if isinstance(item, dict) else item
 
 
-def get_fields(mapping: Optional[Union[Dict[str, Any], str]]) -> List[Dict[str, Any]]:
-    return [{"column": str(next(iter(item.keys()))),
-             "key": str(next(iter(item.values())))}
-            if isinstance(item, dict) else {"column": item, "key": item} for item in mapping] if mapping else []
+def get_target_column(item: Union[Dict[str, str], str]) -> str:
+    return next(iter(item.keys())) if isinstance(item, dict) else item
 
 
-def generate_upsert_stmt(
-        table: str, primary_keys: List[Dict[str, Any]],
-        mapping_fields: List[Dict[str, Any]],
-        db_type: str) -> Any:
+def get_column_mapping(mapping: List[Union[Dict[str, str], str]]) -> List[Dict[str, str]]:
+    return [{"column": next(iter(item.keys())), "key": next(iter(item.values()))} if isinstance(item, dict)
+            else {"column": item, "key": item} for item in mapping] if mapping else []
+
+
+def generate_upsert_stmt(table: Table, primary_keys: List[Dict[str, Any]], db_type: str) -> Any:
     if db_type.lower() == "postgresql":
-        update_fields = [f"{field['column']} = {sa.bindparam(field['key'])}" for field in mapping_fields]
+        for field in primary_keys:
+            if not field["column"] in table.columns:
+                raise ValueError(f"{field['column']} column does not exist in {table.fullname} table")
 
-        insert_fields = ", ".join([field["column"] for field in mapping_fields])
-        pk_fields = ", ".join([field["column"] for field in primary_keys])
+        insert_stmt = insert(table)
 
-        insert_bind_params = ", ".join([f"{sa.bindparam(field['key'])}" for field in mapping_fields])
+        do_update_stmt = insert_stmt .on_conflict_do_update(
+            index_elements=[table.columns[field["column"]] for field in primary_keys],
+            set_=insert_stmt.excluded)
```

Comment:
why is the variable "set_"?

Reply:
See here:
https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#insert-on-conflict-upsert

This is the set clause in the `DO UPDATE SET`